### PR TITLE
Add unit semantic convention generator

### PR DESCRIPTION
**Pull request by justinfoote**
This PR adds a new semantic convention type: units. 
There will only be the one table of units in the spec, but because we're expecting every language to create common constants for metric instrument units (see: [#1177](https://github.com/open-telemetry/opentelemetry-specification/pull/1177), I'd like to make it as easy as possible to generate the constants in code.

The unit yaml looks something like this:
```yaml
groups:
  - id: units
    type: units 
    brief: Specification of commonly used units.
    members:
    - id: percent
      brief: fraction of a total
      value: "%"
    - id: nanosecond
      brief: time
      value: NS
    - id: connections
      brief: connections
      value: "{connections}"
```

And the resulting markdown looks like this:

| Name        | Kind of Quantity         | Unit String   |
| ------------| ----------------         | -----------   |
| percent | fraction of a total | `%` |
| nanosecond | time | `NS` |
| connections | connections | `{connections}` |

There are a few concepts to check out:
* I've created four new classes to define out (currently four) semantic convention types (span, resource, metric, unit). These classes share yaml validation logic, but they each define their own allowed/required keys. The SemanticConvention base class acts as a factory for the four subclasses.
* I removed the `@dataclass` annotation from SemanticConvention, and added a custom constructor that takes a single yaml node as an argument. 

I've added tests for the new functionality, and I've manually run the markdown generator against the opentelemetry-specification repo, and it doesn't have any side-effects.

### CODE REVIEW

File: semantic-conventions/src/opentelemetry/semconv/model/semantic_convention.py
Comment (justinfoote) [inline semantic-conventions/src/opentelemetry/semconv/model/semantic_convention.py:129]: These lines appear to have gotten much simpler, but the complexity is moved from here to the constructor for the class that's returned from the `parse_semantic_convention_type` function. 

This is because the required parameters for each type of convention is different, and passing a single yaml node to the constructor allows the subclasses to own their unique implementations.

File: semantic-conventions/src/opentelemetry/semconv/model/semantic_convention.py
Comment (justinfoote) [inline semantic-conventions/src/opentelemetry/semconv/model/semantic_convention.py:412]: This is added so we only try to sort attributes for semantic conventions that actually have attributes (spans and resources).

File: semantic-conventions/src/opentelemetry/semconv/model/utils.py
Comment (justinfoote) [inline semantic-conventions/src/opentelemetry/semconv/model/utils.py:53]: This is the base class for SemanticConvention (and also UnitMember). In the future, I expect it to be the base class for MetricLabel. 
It owns validation of keys (that all the mandatory keys are present in a yaml node, and that only allowed keys are present), and the validation of values (that the values of the constructed object are allowed).

It may be worth making SemanticAttribute derive from this class, but I didn't do that work now.

File: semantic-conventions/src/tests/semconv/templating/test_markdown.py
Comment (justinfoote) [inline semantic-conventions/src/tests/semconv/templating/test_markdown.py:39]: I _almost_ refactored this whole class, but I settled for just using pytest style assertions, which give much more usable messages (in my opinion). Luckily, it plays well with unittest. :)

**Review by justinfoote — COMMENTED** [review]
@thisthat - I would especially your feedback on this PR.  I see that you and I don't have similar styles, but I tried to match the style of the project at least a little. 
I have some future changes in mind for the metrics work.

File: semantic-conventions/src/opentelemetry/semconv/model/semantic_convention.py
Comment (thisthat) [inline semantic-conventions/src/opentelemetry/semconv/model/semantic_convention.py:83]: Although it is improbable that users will use the `:>` notation, I would still keep the `stip` to avoid newlines in the prefix.
```suggestion
        self.prefix = group.get("prefix", "").strip()
```

File: semantic-conventions/src/opentelemetry/semconv/model/semantic_convention.py
Comment (thisthat) [inline semantic-conventions/src/opentelemetry/semconv/model/semantic_convention.py:91]: This should never be called since we expect subclasses to overwrite this method, right?
If so, I would throw an exception to indicate that something fishy is happening.

**Review by thisthat — APPROVED** [review]
Thank you very much for your great contribution! :rocket: 
I noticed that we have different styles but I believe the more we work on this tool, the faster we will converge to a common one :)

File: semantic-conventions/src/opentelemetry/semconv/model/semantic_convention.py
Comment (Oberon00) [reply semantic-conventions/src/opentelemetry/semconv/model/semantic_convention.py:91]: If you intend to go this way, consider using https://docs.python.org/3/library/abc.html#abc.abstractmethod and the abc module.

Comment (justinfoote) [reply semantic-conventions/src/opentelemetry/semconv/model/semantic_convention.py:91]: This method is used downstream in the SemanticConventionSet to sort attributes. That refactor felt a little too big and rabbit-holey for this PR, and there's a proposal ([#1113](https://github.com/open-telemetry/opentelemetry-specification/issues/1113)) in flight to standardize on "attributes" for metrics as well, so I thought we mightend up undoing the refactor anyway in the very near future.

File: semantic-conventions/src/opentelemetry/semconv/model/utils.py
Comment (aabmass) [inline semantic-conventions/src/opentelemetry/semconv/model/utils.py:56]: nit consider immutable tuples

File: semantic-conventions/src/opentelemetry/semconv/model/semantic_convention.py
Comment (aabmass) [inline semantic-conventions/src/opentelemetry/semconv/model/semantic_convention.py:79]:
```suggestion
        super().__init__(group)
```

File: semantic-conventions/src/opentelemetry/semconv/model/semantic_convention.py
Comment (aabmass) [inline semantic-conventions/src/opentelemetry/semconv/model/semantic_convention.py:268]:
```suggestion
        super().__init__(group)
```

File: semantic-conventions/src/opentelemetry/semconv/model/semantic_convention.py
Comment (aabmass) [inline semantic-conventions/src/opentelemetry/semconv/model/semantic_convention.py:127]: Should the convention class validate itself in the constructor so you don't have to remember to validate separately?

File: semantic-conventions/src/opentelemetry/semconv/model/semantic_convention.py
Comment (aabmass) [inline semantic-conventions/src/opentelemetry/semconv/model/semantic_convention.py:236]:
```suggestion
        super().__init__(group)
```

File: semantic-conventions/src/opentelemetry/semconv/model/semantic_convention.py
Comment (aabmass) [inline semantic-conventions/src/opentelemetry/semconv/model/semantic_convention.py:254]:
```suggestion
        super().__init__(group)
```

File: semantic-conventions/src/opentelemetry/semconv/model/unit_member.py
Comment (aabmass) [inline semantic-conventions/src/opentelemetry/semconv/model/unit_member.py:10]:
```suggestion
        super().__init__(node)
```

**Review by aabmass — APPROVED** [review]
💯

File: semantic-conventions/src/opentelemetry/semconv/model/semantic_convention.py
Comment (justinfoote) [reply semantic-conventions/src/opentelemetry/semconv/model/semantic_convention.py:127]: I considered this. The logical place to put this would be in `ValidatableYamlNode.__init__`, because that superclass owns the validation logic. **..._but_**, the subclasses are likely adding attributes in their own init methods, so the correct place to call it would be _after_ the superclass and subclass init methods had completed.

File: semantic-conventions/src/opentelemetry/semconv/model/semantic_convention.py
Comment (justinfoote) [reply semantic-conventions/src/opentelemetry/semconv/model/semantic_convention.py:55]: I prefer to have this map return something that can be used (like the correct class for the semantic convention). I don't see the value in having the bare strings mapped to integers.

File: semantic-conventions/src/opentelemetry/semconv/model/semantic_convention.py
Comment (justinfoote) [reply semantic-conventions/src/opentelemetry/semconv/model/semantic_convention.py:79]: :exploding_head: This is amazing! I haven't used python professionally since python 2.7 was the thing, and everyone was on the fence about migrating to python 3, so I'm out of the loop on all the new fancy features.

File: semantic-conventions/src/opentelemetry/semconv/model/semantic_convention.py
Comment (aabmass) [reply semantic-conventions/src/opentelemetry/semconv/model/semantic_convention.py:127]: maybe in a factory then?

Comment (justinfoote) [reply semantic-conventions/src/opentelemetry/semconv/model/semantic_convention.py]: I moved the logic to a factory function.
This triggered some renaming (so I wouldn't have a naming collision between `def SemanticConvention` and `class SemanticConvention`), which had some cascading effects.  
I removed several `import SemanticConvention` lines where we were only using the imported class for type declarations, and I refactored references to `SemanticConvention.parse` to a new `parse_semantic_convention_groups` function (and in the process I split the parsing concern away from the data struct concern).

**Review by Oberon00 — CHANGES_REQUESTED** [review]
Please don't remove the enum. The integer values don't matter, but the enum is safer against typos. Even if we don't use static checking like pylint or mypy (yet), we at least would get a runtime error if we type `SemanticConventionType.SAN` whereas with `== "san"` we would just get `False`.

Also the enum is a nice documentation of the possible types.

File: semantic-conventions/src/opentelemetry/semconv/model/semantic_convention.py
Comment (Oberon00) [reply semantic-conventions/src/opentelemetry/semconv/model/semantic_convention.py:55]: Posted my comment about this in the wrong place, sorry: https://github.com/open-telemetry/build-tools/pull/21#pullrequestreview-526042782

> Please don't remove the `SemanticConventionType` enum. The integer values don't matter, but the enum is safer against typos. Even if we don't use static checking like pylint or mypy (yet), we at least would get a runtime error if we type `SemanticConventionType.SAN` whereas with `== "san"` we would just get `False`.

> Also the enum is a nice documentation of the possible types.

File: semantic-conventions/src/opentelemetry/semconv/model/semantic_convention.py
Comment (Oberon00) [inline semantic-conventions/src/opentelemetry/semconv/model/semantic_convention.py:104]: Typo

```suggestion
    # Also, validate that the value of the fields is acceptable
```

File: semantic-conventions/src/opentelemetry/semconv/model/semantic_convention.py
Comment (Oberon00) [inline semantic-conventions/src/opentelemetry/semconv/model/semantic_convention.py:182]: The name TYPE_VALUE is unclear to me, maybe GROUP_TYPE_NAME (a group of this type has this name). Although that also sounds confusing

**Review by aabmass — APPROVED** [review]

File: semantic-conventions/src/opentelemetry/semconv/model/semantic_convention.py
Comment (justinfoote) [reply semantic-conventions/src/opentelemetry/semconv/model/semantic_convention.py]: :+1:

**Comment by Oberon00** [comment]
I think this PR is too huge to properly review it. It seems to be almost a complete rewrite. If others are fine with still merging this, I won't object, personally I would be much happier with smaller pieces if possible (e.g. only have the base class refactoring in a PR, without the addition of Unit, command line interface changes, etc).

But I think we should decide on this PR soon, as having a huge refactoring looming might discourage any other changes, as they would lead to merge conflicts.

**Review by jmacd — APPROVED** [review]

**Comment by jmacd** [comment]
I can't really review this either, but would be happy to see it merged presuming the output is good and others who have looked at it agree.

**Comment by arminru** [comment]
Same as @jmacd.

Both @aabmass and @thisthat already reviewed and approved it - thanks for that! If no one objects by tomorrow, let's merge it.

cc @open-telemetry/specs-approvers

**Comment by justinfoote** [comment]
@open-telemetry/specs-approvers As @arminru suggested above, let's go ahead and merge.

**Comment by arminru** [comment]
Thanks for your contribution and for sticking it out, @justinfoote!

**Pull request merged**
